### ISSUE TITLE
Improve output-format system - Add export feeds

### ISSUE THREAD

**Issue by Makman2** [coala/coala#4815]
Currently the CLI code is generally very complicated. One part is the handling of different output formats, like JSON, formatted or normal output. Everywhere is specific code for different operation modes of coala (showing bears vs normal execution), and this will continue to become more complicated with more features we add.

Inspired by `scrapy`'s export feeds, we can add our own `ExportFeed` module. This module is responsible for accepting objects from everywhere in code (e.g. with a function `export`, or maybe by overriding `print` itself) and print them accordingly to the user.

One question is maybe: Why not just letting coala return the values and we just print them?
There's a major problem with that: For example, we want to prompt results immediately in normal execution mode (checking files) as soon as they arrive. Returning results after the execution has finished is too late. So we need a system that is capable of immediately handling objects to print/export as soon as they arrive.

There are still some major points left to design (what degree of interaction shall happen in the components and in the export feed? Return values of coala command for in-Python-use or only use feed export module?), but this is a basic idea on how to possibly centralize and modularize things again.

Suggestions? :)

**Comment by Makman2** [comment]
CC @shreyans800755 @jayvdb

**Comment by Makman2** [comment]
Yeah in this case your suggestion is super easy to do and will cover everything we would need right now. If that works smoothly, we don't need export feeds at all ;)

**Comment by shreyans800755** [comment]
Sorry, I don't have much knowledge about export feeds.
Can it solve the problem of merging non-interactive and interactive output ? If so, IMO we should go with that.
It will simplify our code and make it much easier to read and extend. Otherwise, we might encounter similar problem while extending code base and need to merge these two functionalities and it would be much more difficult to do it as code expands.

**Comment by Makman2** [comment]
Export feeds are quite simple in fact, in fact it's like logging with custom handler callbacks, but more generally designed to print arbitrary objects intended for use as data. So you should be able to call something like `export()` anywhere from code where you like to export data in some way. In our case I think of specific objects, which implement methods like `__json__` or `__formatted__`. These are called in the export feed handlers accordingly depending on the configuration (like when we specify `--json` in coala, we configure the export feed output-generator to write json)

> If so, IMO we should go with that.
It will simplify our code and make it much easier to read and extend. Otherwise, we might encounter similar problem while extending code base in future and need to merge these two functionalities and it would be much more difficult to do it as code expands.

It would be more flexible, but way more difficult to implement in contrast to your solution which covers all we need right now. But we could do export feeds to further improve our output concept.
